### PR TITLE
Rework lock_core / timers

### PR DESCRIPTION
**Pull request by kilograham**
- Add recursive_mutex
  - Make all locking primitives and sleep use common overridable wait/notify support to allow RTOS
    implementations to replace WFE/SEV with something more appropriate
  - Add busy_wait_ms

### CODE REVIEW

**Review by liamfraser — APPROVED** [review]
Looks okay to me. I spent a while reviewing as requested by Graham.

File: src/common/pico_sync/include/pico/lock_core.h [hunk 1]
```diff
@@ -8,26 +8,183 @@
...
+/** \file lock_core.h
+ *  \defgroup lock_core lock_core
+ *  \ingroup pico_sync
+ * \brief base synchronization/lock primitive support
+ *
+ * Most of the pico_sync locking primitives contain a lock_core_t structure member. This currently just holds a spin
+ * lock which is used only to protect the contents of the rest of the structure as part of implementing the synchronization
+ * primitive. As such, the spin_lock member of lock core is never still held on return from any function for the primitive.
+ *
+ * \ref critical_section is an exceptional case in that it does not have a lock_core_t and simply wraps a pin lock, providing
```
Comment (lurch) [inline src/common/pico_sync/include/pico/lock_core.h:23]: "wraps a pin lock" -> "wraps a spin lock" ?

File: src/common/pico_sync/include/pico/lock_core.h [hunk 2]
```diff
@@ -8,26 +8,183 @@
...
+ * methods to lock and unlock said spin lock.
+ *
+ * lock_core based structures work by locking the spin lock, checking state, and then deciding whether they additionally need to block
+ * or notify when the spin lock is released. In the blocking case, they will wake up again in the future, and try the process again.
+ *
+ * By default the SDK just uses the processors' events via SEV and WEV for notification and blocking as these are sufficient for
+ * cross core, and notification from interrupt handlers. However macros are defined in this file that abstract the wait
+ * and notify mechanisms to allow the SDK locking functions to effectively be used within an RTOS on other environment.
```
Comment (lurch) [inline src/common/pico_sync/include/pico/lock_core.h:31]: "RTOS on other environment" -> "RTOS or other environment" ?

File: src/common/pico_sync/include/pico/lock_core.h [hunk 3]
```diff
@@ -8,26 +8,183 @@
...
+ *
+ * When implementing an RTOS, it is desirable for the SDK synchronization primitives that wait, to block the calling task (and immediately yield),
+ * and those that notify, to wake a blocked task which isn't on processor. At least the wait macro implementation needs to be atomic with the protecting
+ * spin_lock unlock from the callers point of view; i.e. the task should unlock the spin lock when as it starts its wait. Such implementation is
```
Comment (lurch) [inline src/common/pico_sync/include/pico/lock_core.h:35]: "spin lock when as it starts" -> "spin lock when it starts" ?

File: src/common/pico_sync/include/pico/lock_core.h [hunk 4]
```diff
@@ -8,26 +8,183 @@
...
+ * a missed notification). In other words this method should always wake up in response to a lock_internal_spin_unlock_with_notify
+ * for the same lock, which completes after this call starts.
+ *
+ * In an ideal implementation, this method would return exactly after the corresponding lock_internal_spin_unlock_with_notify
+ * has subsequently been called on the same lock instance, however this method is free to return at _any_ point before that;
+ * this macro is _always_ used in a loop which locks the spin lock, checks the internal locking primitive state and then
+ * waits again if the calling thread should not proceed.
+ *
+ * By default this macro simply unlocks the spin lock, and then performs a WFE, but may be overridden
+ * (e.g. to actually block the RTOS task).
```
Comment (lurch) [inline src/common/pico_sync/include/pico/lock_core.h:113]: Mixture of using "this method" and "this macro" (although I dunno if this difference is significant?)

File: src/common/pico_sync/include/pico/lock_core.h [hunk 5]
```diff
@@ -8,26 +8,183 @@
...
+#endif
+
+#ifndef sync_internal_yield_until_before
+/*! \brief   yield to other processing until some time before the requested time
+ *  \ingroup lock_core
+ *
+ * This method is provided for cases where the caller has no useful work to do
+ * until the specified time.
+ *
+ * By default this method does nothing, however if can be overridden (for example by an
```
Comment (lurch) [inline src/common/pico_sync/include/pico/lock_core.h:181]: "if can" -> "it can"

File: src/common/pico_sync/include/pico/mutex.h [hunk 6]
```diff
@@ -28,19 +28,33 @@ extern "C" {
...
     lock_core_t core;
-    int8_t owner; //! core number or -1 for unowned
+    lock_owner_id_t owner;      //! owner id LOCK_INVALID_OWNER_ID for unowned
+    uint8_t recursion_state;    //! 0 means non recursive (owner or unowned)
+                                //! 1 is a maxed out recursive lock
+                                //! 2-254 is an owned lock
+                                //! 255 is an un-owned lock
 } mutex_t;
 
+#define MAX_RECURSION_STATE ((uint8_t)255)
```
Comment (lurch) [inline src/common/pico_sync/include/pico/mutex.h:40]: Hopefully people won't accidentally assume that `MAX_RECURSION_STATE` corresponds to the "1 is a maxed out recursive lock" described above?

File: src/common/pico_sync/include/pico/mutex.h [hunk 7]
```diff
@@ -28,19 +28,33 @@ extern "C" {
  *
  * See \ref critical_section.h for protecting access between multiple cores AND IRQ handlers
  */
-
 typedef struct __packed_aligned mutex {
     lock_core_t core;
-    int8_t owner; //! core number or -1 for unowned
+    lock_owner_id_t owner;      //! owner id LOCK_INVALID_OWNER_ID for unowned
+    uint8_t recursion_state;    //! 0 means non recursive (owner or unowned)
```
Comment (lurch) [inline src/common/pico_sync/include/pico/mutex.h:34]: I wonder if it'd have been worth making this `uint8_t` a typedef (e.g. `recursion_state_t`) ? :shrug:   Or perhaps we already have enough typedefs scattered throughout the place.

**Pull request merged**